Perl_sv_clear: unrolled SvREFCNT_dec and sv_free2....isn't that?!

**Description**
Towards the bottom of `Perl_sv_clear`, which is frequently a very hot function, there is the following comment:
`/* unrolled SvREFCNT_dec and sv_free2 follows: */`

Sounds good, makes sense that it would unroll `SvREFCNT_dec` and `sv_free2`.
But that's not actually what it does:

```
            /* unrolled SvREFCNT_dec and sv_free2 follows: */

            if (!sv)
                continue;
            if (!SvREFCNT(sv)) {
                sv_free(sv);
                continue;
            }
            .....
```
At least nowadays, `sv_free` means calling `Perl_sv_free`:

```
void
Perl_sv_free(pTHX_ SV *const sv)
{
    SvREFCNT_dec(sv);
}
```
`SvREFCNT_dec` is an inline function in `sv_inline.h`:
```
PERL_STATIC_INLINE void
Perl_SvREFCNT_dec(pTHX_ SV *sv)
{
    if (LIKELY(sv != NULL)) {
        U32 rc = SvREFCNT(sv);
        if (LIKELY(rc > 1))
            SvREFCNT(sv) = rc - 1;
        else
            Perl_sv_free2(aTHX_ sv, rc);
    }
}
```
So instead of unrolling `SvREFCNT_dec` and `sv_free2`, it calls a function (probably inlined)
to call `SvREFCNT_dec` which likely will call `sv_free2`. That's not unrolled at all!

The likely impact of this is some SV freeing is taking longer than it strictly has to. Looking at
gcov coverage when running the test harness, there are 35099459 calls to `Perl_sv_free` and
797851111 calls to `Perl_sv_free2`, so maybe ~4% of cleared SVs. 

**Expected behavior**
I haven't sat down to figure this out. It seems like the status quo in `Perl_sv_clear` must be 
wrong though and either the comment should be amended, (more likely) the call to `sv_free`
is intended to be a recursive call back to `sv_free2`, or some other code change should happen.

Besides `Perl_sv_clear`, only _ext/Opcode/Opcode.xs_ and _dist/Storable/Storable.xs_ seem
to call `sv_free` directly. Probably they should be using `SvREFCNT_dec` or calling `sv_free2`
instead.

**Perl configuration**
blead

> I really want this issue to just be about the code following `/* unrolled SvREFCNT_dec and sv_free2 follows: */`. Attempting to shoehorn in critiques about how the core does memory allocations or uses the tmps stack, or how the XS or C APIs work, or what gmake does is unlikely to result in changes to those things. More likely it will just put off engagement with this specific issue.
> 
> To achieve changes on those other topics, focussed Issues or Pull Requests (or topics on the mailing list) seem like better vehicles to me.
> 

+1

> FWIW This builds and the test harness passes, but it might not be the best way to update this call site:
> 
> ```
> diff --git a/sv.c b/sv.c
> index f0b0b006d4..b7c950506f 100644
> --- a/sv.c
> +++ b/sv.c
> @@ -7125,7 +7125,7 @@ Perl_sv_clear(pTHX_ SV *const orig_sv)
>              if (!sv)
>                  continue;
>              if (!SvREFCNT(sv)) {
> -                sv_free(sv);
> +                Perl_sv_free2(sv, 0);
>                  continue;
>              }
>              if (--(SvREFCNT(sv)))
> ```

For me this passed on an unthreaded build on Linux using both `gcc` and `clang` as the compiler -- but `make` failed on *threaded* builds with both `gcc` and `clang` -- albeit with different error output -- on the same platform.

Tail of `make` output:
```
$ clang --version
Ubuntu clang version 18.1.3 (1ubuntu1)
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin

$ sh ./Configure -des -Dusedevel -Duseithreads -Dcc=clang && make test_harness
...
clang -c -DPERL_CORE -D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -std=c99 -O2 -Wall -Werror=pointer-arith -Werror=vla -Wextra -Wno-long-long -Wno-declaration-after-statement -Wc++-compat -Wwrite-strings sv.c
sv.c:7128:36: error: too few arguments to function call, expected 3, have 2
 7128 |                 Perl_sv_free2(sv, 0);
      |                 ~~~~~~~~~~~~~      ^
./proto.h:4701:1: note: 'Perl_sv_free2' declared here
 4701 | Perl_sv_free2(pTHX_ SV * const sv, const U32 refcnt);
      | ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
make: *** [makefile:265: sv.o] Error 1
```

```
$ gcc --version
gcc (Ubuntu 13.2.0-23ubuntu4) 13.2.0

$ sh ./Configure -des -Dusedevel -Duseithreads -Dcc=gcc && make test_harness
...
gcc -c -DPERL_CORE -D_REENTRANT -D_GNU_SOURCE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -std=c99 -O2 -Wall -Werror=pointer-arith -Werror=vla -Wextra -Wno-long-long -Wno-declaration-after-statement -Wc++-compat -Wwrite-strings -Wno-use-after-free sv.c
sv.c: In function ‘Perl_sv_clear’:
sv.c:7128:31: warning: passing argument 1 of ‘Perl_sv_free2’ from incompatible pointer type [-Wincompatible-pointer-types]
 7128 |                 Perl_sv_free2(sv, 0);
      |                               ^~
      |                               |
      |                               SV * {aka struct sv *}
In file included from sv.c:32:
perl.h:224:22: note: expected ‘PerlInterpreter *’ {aka ‘struct interpreter *’} but argument is of type ‘SV *’ {aka ‘struct sv *’}
  224 | #  define pTHX  tTHX my_perl PERL_UNUSED_DECL
      |                      ^
perl.h:229:25: note: in expansion of macro ‘pTHX’
  229 | #  define pTHX_         pTHX,
      |                         ^~~~
proto.h:4701:15: note: in expansion of macro ‘pTHX_’
 4701 | Perl_sv_free2(pTHX_ SV * const sv, const U32 refcnt);
      |               ^~~~~
sv.c:7128:17: error: too few arguments to function ‘Perl_sv_free2’
 7128 |                 Perl_sv_free2(sv, 0);
      |                 ^~~~~~~~~~~~~
In file included from perl.h:6160:
proto.h:4701:1: note: declared here
 4701 | Perl_sv_free2(pTHX_ SV * const sv, const U32 refcnt);
      | ^~~~~~~~~~~~~
make: *** [makefile:265: sv.o] Error 1
```

--- a/sv.c
+++ b/sv.c
@@ -7120,12 +7120,18 @@ Perl_sv_clear(pTHX_ SV *const orig_sv)
                 }
             }
 
-            /* unrolled SvREFCNT_dec and sv_free2 follows: */
+            /* Do the equivalent of SvREFCNT_dec(sv), except:
+             - for the case of RC==1, inline the actions normally taken
+               by sv_free2() prior it calling sv_clear(), and handle the
+               sv_clear() actions ourselves (without needing to
+               recurse).
+             - For the exceptional case of RC==0, do a traditional
+               recursive free. */
 
             if (!sv)
                 continue;
             if (!SvREFCNT(sv)) {
-                sv_free(sv);
+                Perl_sv_free2(aTHX_ sv, 0);
                 continue;
             }
             if (--(SvREFCNT(sv)))

